stl/inc/bit

```diff
@@ -94,18 +96,117 @@ _NODISCARD constexpr _Ty rotr(const _Ty _Val, const int _Rotation) noexcept {
+    const bool _Have_lzcnt = __isa_available >= __ISA_AVAILABLE_AVX2;
+    constexpr int _Digits  = numeric_limits<_Ty>::digits;
+    // lzcnt (when it doesn't fall back to bsr) is defined correctly for zero
+    // bsr has undefined output for zero
+    if (!_Have_lzcnt && _Val == 0) {
+        return _Digits;
+    }
+    // We use lzcnt (actually bsr if lzcnt is not supported) now that we know
+    // we're not zero. We can do this because lzcnt and bsr share the same instruction
+    // encoding.
+    if constexpr (_Digits <= 16) {
+        return static_cast<int>(__lzcnt16(_Val) - (16 - _Digits));
+    } else if constexpr (_Digits == 32) {
+        return static_cast<int>(__lzcnt(_Val));
+    } else {
+#ifdef _M_IX86
+        const unsigned int _High = _Val >> 32;
+        const auto _Low          = static_cast<unsigned int>(_Val);
+        int _Result              = _Checked_x86_x64_countl_zero(_High);
+        _Result += _High == 0 ? _Checked_x86_x64_countl_zero(_Low) : 0;
+        return _Result;
+#else // ^^^ _M_IX86 / !_M_IX86 vvv
+        return static_cast<int>(__lzcnt64(_Val));
+#endif // _M_IX86
+    }
+    // note: we don't need to call a fallback here because
+    // all supported x86 processors at least have bsr/bsf
+}
+
+template <class _Ty>
+_NODISCARD int _Checked_x86_x64_popcount(const _Ty _Val) noexcept {
+    const bool _Have_popcnt = __isa_available >= __ISA_AVAILABLE_SSE42;
+    constexpr int _Digits   = numeric_limits<_Ty>::digits;
+    if (_Have_popcnt) {
+        if constexpr (_Digits <= 16) {
+            return static_cast<int>(__popcnt16(_Val));
+        } else if constexpr (_Digits == 32) {
+            return static_cast<int>(__popcnt(_Val));
+        } else {
+#ifdef _M_IX86
+            return static_cast<int>(__popcnt(_Val >> 32) + __popcnt(static_cast<unsigned int>(_Val)));
+#else // ^^^ _M_IX86 / !_M_IX86 vvv
+            return static_cast<int>(__popcnt64(_Val));
+#endif // _M_IX86
+        }
+    } else {
+        return _Popcount_helper(_Val);
+    }
+}
+#endif // defined(_M_IX86) || defined(_M_X64)
+
+
+#if defined(_M_ARM) || defined(_M_ARM64)
+template <class _Ty>
+_NODISCARD int _Checked_arm_arm64_countl_zero(const _Ty _Val) noexcept {
     constexpr int _Digits = numeric_limits<_Ty>::digits;
     if (_Val == 0) {
         return _Digits;
     }
 
-    if constexpr (sizeof(_Ty) <= sizeof(unsigned int)) {
-        return __builtin_clz(_Val) - (numeric_limits<unsigned int>::digits - _Digits);
+    if constexpr (_Digits <= 32) {
+        return _CountLeadingZeros(_Val);
     } else {
-        return __builtin_clzll(_Val) - (numeric_limits<unsigned long long>::digits - _Digits);
+        return _CountLeadingZeros64(_Val);
+    }
+}
+#endif // defined(_M_ARM) || defined(_M_ARM64)
+
+template <class _Ty, enable_if_t<_Is_standard_unsigned_integer<_Ty>, int> _Enabled>
+_NODISCARD constexpr int countl_zero(const _Ty _Val) noexcept {
+    if (!_STD is_constant_evaluated()) {

```

Comment:
Suggestion: This test could be reversed. (We conventionally prefer to avoid negated tests, when doing so results in reasonable code.) In this case, I believe it would significantly improve readability, since it would read as "if we're constant evaluated, call constexpr helper. Otherwise, preprocessor architecture-dependent logic", and that would be separated from the constexpr branch.

Reply:
Shouldn't it be `_CONSTEXPR_IF` ?

Reply:
That's a good question. `<bit>` is C++20, so it can unconditionally assume that `if constexpr` is available. However, `if constexpr` must never be used with `is_constant_evaluated()` (surprisingly!). That's because `is_constant_evaluated()` reports whether it's being called in a "manifestly constant-evaluated" context, and `if constexpr` is such a context (see [[expr.const]/14.2](https://eel.is/c++draft/expr.const#14.2)), so it would always report `true`. Some compilers (e.g. Clang) will warn about this.

That's why we need to use plain `if` to test `is_constant_evaluated()`.

Reply:
This is not yet resolved.